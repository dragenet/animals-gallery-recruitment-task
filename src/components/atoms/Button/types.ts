import { ButtonHTMLAttributes } from 'react';

export type ButtonBaseProps<T> = Pick<
  ButtonHTMLAttributes<T>,
  'onClick' | 'children' | 'className'
>;
export interface ButtonProps<T> extends ButtonBaseProps<T> {
  variant?: ButtonVariant;
}

export type ButtonVariant = 'primary' | 'secondary';
