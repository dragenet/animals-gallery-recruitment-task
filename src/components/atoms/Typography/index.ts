export * from './types';
export * from './Typography';
export * from './utils';
