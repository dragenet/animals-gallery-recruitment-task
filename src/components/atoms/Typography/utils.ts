import { css, ThemedStyledProps } from 'styled-components';

import { Colors, FontSize } from '../../../theme';
import { TypographyProps } from './types';

export const fontSize = (fontSize: FontSize) => css`
  font-size: ${({ theme }) => theme.fontSizes[fontSize]};
`;

export const color = (color: Colors = 'black') => css`
  color: ${(props) => props.theme.colors[color]};
`;

export const bold = css`
  font-weight: bold;
`;

export const typographyMixin =
  (options?: Partial<TypographyProps>) =>
  <T extends TypographyProps, V>(props: ThemedStyledProps<T, V>) => {
    const currentProps = { ...options, ...props };

    return css`
      ${currentProps.fontSize && fontSize(currentProps.fontSize)}
      ${currentProps.color && color(currentProps.color)}
  ${currentProps.bold && bold}
    `;
  };
