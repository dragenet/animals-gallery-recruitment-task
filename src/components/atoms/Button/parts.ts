import styled, { css } from 'styled-components';

import { color, typographyMixin } from '~/components/atoms/Typography';

import { ButtonProps } from './types';

export const ButtonBase = styled.button<ButtonProps<unknown>>`
  cursor: pointer;
  padding: 1em;
  border-radius: 0.5em;
  ${typographyMixin({ bold: true })}
`;

const primaryButtonCss = css`
  border: none;
  background-color: ${({ theme }) => theme.colors.indigo};
  ${color('white')}
`;

const secondaryButtonCss = css`
  border: 1px solid ${({ theme }) => theme.colors.indigo};
  background-color: ${({ theme }) => theme.colors.white};
  ${color('indigo')}
`;

export const ButtonPrimary = styled(ButtonBase)`
  ${primaryButtonCss}

  &:hover {
    ${secondaryButtonCss}
  }
`;

export const ButtonSecondary = styled(ButtonBase)`
  ${secondaryButtonCss}

  &:hover {
    ${primaryButtonCss}
  }
`;
