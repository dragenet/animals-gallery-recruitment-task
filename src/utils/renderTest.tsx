import { render } from '@testing-library/react';

import { RootProvider } from '~/components/atoms/RootProvider';

export const renderTest = (children: JSX.Element) =>
  render(<RootProvider>{children}</RootProvider>);
