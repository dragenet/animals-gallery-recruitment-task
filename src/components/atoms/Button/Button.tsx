import { buttonVariantMapping } from '~/components/atoms/Button/constants';

import { ButtonProps } from './types';

export const Button = <T,>({ children, variant = 'primary', ...props }: ButtonProps<T>) => {
  const Button = buttonVariantMapping[variant];
  return (
    <>
      <Button {...props}>{children}</Button>
    </>
  );
};
