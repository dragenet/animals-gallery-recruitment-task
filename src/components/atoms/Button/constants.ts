import { FC } from 'react';

import { ButtonPrimary, ButtonSecondary } from '~/components/atoms/Button/parts';
import { ButtonBaseProps, ButtonVariant } from '~/components/atoms/Button/types';

export const buttonVariantMapping: Record<ButtonVariant, FC<ButtonBaseProps<unknown>>> = {
  primary: ButtonPrimary,
  secondary: ButtonSecondary,
};
