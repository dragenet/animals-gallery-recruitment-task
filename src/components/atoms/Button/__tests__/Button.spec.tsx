import { screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { Button } from '~/components/atoms/Button';
import { renderTest } from '~/utils/renderTest';

describe('Button', () => {
  it('Primary renders correctly', () => {
    const { container } = renderTest(<Button>Hello there!</Button>);

    expect(container).toMatchSnapshot();
  });

  it('Secondary renders correctly', () => {
    const { container } = renderTest(<Button variant="secondary">Hello there secondary!</Button>);

    expect(container).toMatchSnapshot();
  });

  it('is clickable', async () => {
    const handleClickSpy = vi.fn();
    const { user } = renderTest(
      <Button variant="secondary" onClick={handleClickSpy}>
        Hello there secondary!
      </Button>,
    );

    await user.click(screen.getByRole('button'));

    expect(handleClickSpy).toHaveBeenCalledOnce();
  });
});
