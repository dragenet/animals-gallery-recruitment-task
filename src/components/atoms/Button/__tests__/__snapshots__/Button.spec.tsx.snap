// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Primary renders correctly 1`] = `
<div>
  <button
    class="sc-hLseeU sc-eDDNvR lmVMho hczWgw"
  >
    Hello there!
  </button>
</div>
`;
